python/ray/serve/grpc_util.py

```diff
@@ -0,0 +1,79 @@
+from typing import Any, Dict, List, Optional, Tuple
+
+import grpc
+
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="beta")
+class RayServegRPCContext:
+    """Context manager to set and get gRPC context.
+
+    This class implements most of the methods from ServicerContext
+    (see: https://grpc.github.io/grpc/python/grpc.html#grpc.ServicerContext) so it's
+    serializable and can pass with the request to be used on the deployment.
+    """
+
+    def __init__(self, grpc_context: grpc._cython.cygrpc._ServicerContext):
+        self._auth_context = grpc_context.auth_context()
+        self._code = grpc_context.code()
+        self._details = grpc_context.details()
+        self._invocation_metadata = [
+            (key, value) for key, value in grpc_context.invocation_metadata()
+        ]
+        self._peer = grpc_context.peer()
+        self._peer_identities = grpc_context.peer_identities()
+        self._peer_identity_key = grpc_context.peer_identity_key()
+        self._trailing_metadata = [
+            (key, value) for key, value in grpc_context.trailing_metadata()
+        ]
+        self._compression = None
+
+    def auth_context(self) -> Dict[str, Any]:
+        return self._auth_context
+
+    def code(self) -> grpc.StatusCode:
+        return self._code
+
+    def details(self) -> str:
+        return self._details
+
+    def invocation_metadata(self) -> List[Tuple[str, str]]:
+        return self._invocation_metadata
+
+    def peer(self) -> str:
+        return self._peer
+
+    def peer_identities(self) -> Optional[bytes]:
+        return self._peer_identities
+
+    def peer_identity_key(self) -> Optional[str]:
+        return self._peer_identity_key
+
+    def trailing_metadata(self) -> List[Tuple[str, str]]:
+        return self._trailing_metadata
+
+    def set_code(self, code: grpc.StatusCode):
+        self._code = code
+
+    def set_compression(self, compression: grpc.Compression):
+        self._compression = compression
+
+    def set_details(self, details: str):
+        self._details = details
+
+    def set_trailing_metadata(self, trailing_metadata: List[Tuple[str, str]]):
+        self._trailing_metadata += trailing_metadata
```

Comment:
The method name is _set_ `trailing_metadata`, but this is additive. Does this match the semantics of grpc?

Reply:
It does not 😅 In gRPC those attributes are override by the latest call before the request is terminated. I want to ensure our request id is still passed together with whatever users set to the trailing metadata so implemented this way.

Reply:
gotcha, will fix it👍

Reply:
Fixed, and also add a test case to ensure request id exist in the response when the user set something to the trailing metadata.